### ISSUE TITLE
If a composer is displayed, I want an autosave of draft emails when writing  it

### ISSUE THREAD

**Issue by fabienmoyon** [OpenPaaS-Suite/esn-frontend-inbox#434]
#### CURRENT

Draft email are save :
- before sending 
- if we click on the save icon (on for the mailto SPA)
- if we click on the cross icon to quite the compose 

#### EXPECTED 

1. I want an autosave of draft emails when writing it : 

- Trigger the save event with a denounce when we stop writing in the body or when we stop adding attendies during more than 3 seconds
- Trigger the save event if the cursor leave the composer during more than 3 seconds

2. If the save to the server is not possible (network error or other) we save the draft email in the local storage to be able to save it later 

If saving to the server is not possible (network error or other), we save the draft email in the local storage. Like this we can save it later and keep the draft email up to date the next time we try to save or the internet connection comes back

#### SPEC 

The save is done by a `setMessages` jmap call to James

**Comment by chibenwa** [comment]
> Trigger the save event with a denounce when we stop writing in the body or when we stop adding attendies during more than 3 seconds

Be sure to make this easily configurable, and desactivable from an admin point of view.

3 seconds seems to be very often to me... I would be worry having thousands of users saving there drafts every 3 seconds...

> If the save to the server is not possible (network error or other) we save the draft email in the local storage to be able to save it later

This sounds just like a better option to me to be honnest...

-------------------------

Are you sure saving the current composition on the backend side is common?

I had a quick look to what GitHub does for comments (like this one) - it gets saved when I refresh, yet there are no interactions with the backend servers...

I would definitly prefer only going for "local storage", keeping draft stuff when the user explicitly wants to save it as a draft.

**Comment by rouazana** [comment]
> Are you sure saving the current composition on the backend side is common?

Roundcube and GMail are doing it at least. If 3s is too fast, maybe 1mn could be fine. BTW I don't think you can compare GitHub to a mail software.

**Comment by chibenwa** [comment]
> BTW I don't think you can compare GitHub to a mail software.

Sure but they do care about "not loosing user input" and the user experience is good.

> Roundcube and GMail are doing it at least. If 3s is too fast, maybe 1mn could be fine. 

Saying "We need to do like the others without thinking" is IMO not the right approach.

Let's speak about the needs. Not loose user input. Do we need to save it so that we can access it on another device? On another protocol? Most of the time when you screw up your input (accidental page reload, network connection PB) you do access it with the same user agent. Hence I qualify those needs of "overkill".

Let's speak about the risks. Thousands of people composer open saving data potentially every 3 seconds seems very scary to me. I'd like to protect my data tier with an approach that scales and push it as a frontend concern.

**Comment by Arsnael** [comment]
I agree that 3s is a door to get flooded on the James side. Gmail is likely doing it, but they likely have all the infra of the world behind to be able to. Not likely what we have :)

I'm not convinced about the cache either. If your computer crash, or if you close your browser (some users clear their browser cache upon closing) it might not result to a great experience. 

But one minute is too long IMO. I would think of something between 15 - 30s?

**Comment by chibenwa** [comment]
> I'm not convinced about the cache either. If your computer crash, or if you close your browser (some users clear their browser cache upon closing) it might not result to a great experience.

Local storage is persistent.

If you computer burns my bet is that the mail you were composing is the least of your problems.

**Comment by Arsnael** [comment]
> If you computer burns my bet is that the mail you were composing is the least of your problems.

Without going to the burning scenario, once again, you can configure your browser to clear your local storage when you close it. And don't think nobody uses it :)

**Comment by chibenwa** [comment]
> you can configure your browser to clear your local storage when you close it

Then you understand the consequences that goes with it!

**Comment by rouazana** [comment]
> Without going to the burning scenario, once again, you can configure your browser to clear your local storage when you close it. And don't think nobody uses it :)

It makes me think of a scenario: you are writing your email in a private session of your browser (because your are in an internet café or more generally on someone's else computer). You lose internet connection / power / a cat suddenly closes the windows. In these cases you'll appreciate to have your draft.

**Comment by tanandy** [comment]
I agree with Raphael this use case if important. Btw, using local storage to store draft is a not a good thing for security concern, especially if we dont encrypt things for ex. Gmail seems to not use service workers neither.
It seems, gmail save draft periodically and under certain conditions but you need to be connected to the internet (correct me if im wrong)

for example: when you create a new message and start typing you can see (Draft saved on your composer), if you do the same beeing offline and close the composer, it seems that it doesnt work

**Comment by fabienmoyon** [comment]
This task is in draft, I will precise use cases with your feedback and with the best solution for us. 

Yes exactly, Gmail trigger the call to save the draft after a no activity in the compose (a debounce when stop writing in the body or when we stop adding attendees).

We can keep the same logic with a debounce of 15-30s WDYT?

I will precise this and update scenarios in the expected section.

**Comment by chibenwa** [comment]
Disabling auto-save MUST anyway be easy.

Configuring a saving period MUST be easy.

This means via **configuration** and **documented**.

Also we MUST plan a backend test campain to better understand the impact.
